### PR TITLE
Add chirp z transform.

### PR DESCRIPTION
**Pull request by ebezzam**
First version of CZT. There is still residual error when using float32 / complex64.

### CODE REVIEW

**Comment by ebezzam** [comment]
I can turn the czt script into proper test, just pushing in case we want to dig deeper into the residual error.

Here's a reference implementation (numpy): https://github.com/imagingofthings/pyFFS/blob/6c72c2b92f347ef186d3f65a143fbe3e95f04ad9/pyffs/czt.py#L85

**Comment by GJBoth** [comment]
@ebezzam I think Dip is handling this mostly but do shout out when you want to merge and I'll give it a last look too!

**Comment by ebezzam** [comment]
Re propagation, it wouldn't affect this code so it can be a separate PR.

**Comment by GJBoth** [comment]
> They have different implementations because cztn has overhead to handle multiple axes, which makes it slower to compute 1D CZT than the czt implementation. I have another profiling script to compare runtimes, which I haven't pushed. In that profiling script I also compare runtime for 2D CZT (applying cztn vs applying czt along each dimension) and in this case cztn is faster.

I'm surpised about the runtime difference. I'd expect the compiler to take care of that, and there shouldn't be a difference between 1D and multiple axes 1D. Are you sure you're timing correctly?

**Comment by ebezzam** [comment]
@GJBoth Given above timings and if we wanted to a single API, one option would be:

- Keep [this multidimensional approach](https://github.com/ebezzam/chromatix/blob/6ac0491916b890b1b75f35693865c41dd453a4a8/src/chromatix/utils/czt.py#L179) (which was actually faster according to above profiling) and rename to `czt`.
- Rename the [1D CZT](https://github.com/ebezzam/chromatix/blob/6ac0491916b890b1b75f35693865c41dd453a4a8/src/chromatix/utils/czt.py#L46) to something like `_czt` as the above multidimensional [calls it](https://github.com/ebezzam/chromatix/blob/6ac0491916b890b1b75f35693865c41dd453a4a8/src/chromatix/utils/czt.py#L185) but it doesn't need to be exposed to users.
- Delete the [cztn](https://github.com/ebezzam/chromatix/blob/6ac0491916b890b1b75f35693865c41dd453a4a8/src/chromatix/utils/czt.py#L93). This one was more gratifying implementation-wise as it used `einsum` and `fftn` but always comes out a bit slower 😅

**Comment by GJBoth** [comment]
Yes, this is what I suspected, and how timings works in Jax. Your first evaluation is longer because it's when the function is compiled, so to get good timings you need to run the function once, and only then do your timings. Once this is done I suspect the difference between `czt` and `cztn` will disappear. That said, I think in the end we should stick with what numpy does and use both a `czt` and `cztn`, just so we don't invent our own stuff :-). That means everything is ready for a final review and to be merged, right?

**Comment by ebezzam** [comment]
@GJBoth Yes it is ready. I just updated `cztn` to the version that performs the looped implementation, i.e. applying `czt` along each dimension as it is faster than the version that applies `einsum` and `fftn` (now called `_cztn`). Do you want to keep this second version of CZT along multiple dimensions?

File: src/chromatix/utils/czt.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    Args:
+        x: Input signal to transform.
+        m: Number of samples in the output.
+        a: The starting point in the complex plane.
+        w: The ratio between points in each step.
+    """
+    axes, a, w = _verify_cztn_input(x, a=[a], w=[w], m=[m], axes=[axis])
+    axis = axes[0]
+    a = a[0]
+    w = w[0]
```
Comment (GJBoth) [inline src/chromatix/utils/czt.py:68]: First you turn it into a list, than unpack again. Is that necessary?

File: src/chromatix/utils/czt.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+    # perform CZT
+    broadcast_shape = [1] * len(x.shape)
+    broadcast_shape[axis] = n
+    y = jnp.fft.fft(x * Awk2.reshape(broadcast_shape), n_czt, axis=axis)
+    broadcast_shape[axis] = n_czt
+    y = jnp.fft.ifft(y * Fwk2.reshape(broadcast_shape), axis=axis)
+    broadcast_shape[axis] = m
+    y = y[idx] * wk2.reshape(broadcast_shape)
+    return y
```
Comment (GJBoth) [inline src/chromatix/utils/czt.py:90]: There's a lot happening here and it's not that clear to me. You're doing a lot of reshaping and slicing which I'm not entirely sure is necessary? Just for reference, I worked with claude to make a version where we move the transformed axis to the back, really simplifying things! I haven't checked this is 100% correct, but it shows how we can make this much clearer, I think:

```
def chirp_z_transform(
    x: ArrayLike,  # Input array of any shape
    M: int,        # Number of output points
    W: complex,    # Complex ratio between points
    Z_0: complex = 1.0,  # Starting point
    *,
    axis: int = -1       # Axis to transform
) -> jnp.ndarray:
    """
    Compute the Chirp Z-transform using JAX.
    
    Args:
        x: Input array to transform
        M: Number of output points
        W: Complex ratio between points on the z-plane
        Z_0: Starting point in the z-plane
        axis: Axis along which to compute the transform
    
    Returns:
        Array with the same shape as input except length M along the transform axis
    """
    x = jnp.moveaxis(x, axis, -1)
    N = x.shape[-1]
    
    n_range = jnp.arange(N)
    k_range = jnp.arange(M)
    
    chirp_n = W**(n_range**2 / 2.0)
    chirp_k = W**(-k_range**2 / 2.0)
    
    nfft = N + M - 1
    H = jnp.zeros(nfft, dtype=complex).at[:M].set(chirp_k)
    H_fft = jnp.fft.fft(H)
    
    y = x * chirp_n
    Y = jnp.fft.fft(y, n=nfft, axis=-1)
    result = jnp.fft.ifft(Y * H_fft, axis=-1)[..., :M] * (Z_0**(-k_range))
    
    return jnp.moveaxis(result, -1, axis)
```

File: src/chromatix/utils/czt.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+def _verify_axes(x, axes):
+    axes = [a + x.ndim if a < 0 else a for a in axes]
+    if any(a >= x.ndim or a < 0 for a in axes):
+        raise ValueError("axes exceeds dimensionality of input")
+    if len(set(axes)) != len(axes):
+        raise ValueError("all axes must be unique")
+    return axes
+
+
+def _verify_cztn_input(x, a, w, m, axes):
```
Comment (GJBoth) [inline src/chromatix/utils/czt.py:16]: This is where we need to make some big changes. Jitting doesn't allow stuff like this as it depends on the incoming values AND isn't functional. Our options are:
1) Not do it at all.
2) Use Jaxtyping: https://github.com/patrick-kidger/jaxtyping
3) Use checkify: https://docs.jax.dev/en/latest/debugging/checkify_guide.html#

I lean more towards taking this all out, and figuring out at a later stage how to do error handling nicely. I know this is good practice in numpy, but it doesn't work in Jax, unfortunately.

**Review by GJBoth — CHANGES_REQUESTED** [review]
Right so we still have some work to do per the comments (mostly adapting to Jax style) but we're almost there :-).

File: pyproject.toml [hunk 4]
```diff
@@ -13,7 +13,7 @@ dependencies = ["jax >= 0.4.1", "einops >= 0.6.0", "flax >= 0.6.3", "chex>=0.1.5
 version = "0.3.0"
 
 [project.optional-dependencies]
-dev = ["mypy>= 0.991", "pytest>=7.2.0", "ruff==0.9.0", "pre-commit>=4.0.0"]
+dev = ["mypy>= 0.991", "pytest>=7.2.0", "ruff==0.9.0", "pre-commit>=4.0.0", "matplotlib>=3.10.0"]
```
[the remark this answers is not in the document]
Comment (ebezzam) [reply pyproject.toml:16]: Not directly related to this PR, but because the example notebooks need matplotlib for plotting. But I'm fine to remove it

File: src/chromatix/utils/czt.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+            )
+        if not jnp.isclose(abs(w[d]), 1):
+            raise ValueError("Parameter[w[d]] must lie on the unit circle.")
+        if m[d] <= 0:
+            raise ValueError("Parameter[m[d]] must be positive.")
+
+    return axes, a, w
+
+
+def czt(x: Array, m: int, a: complex, w: complex, axis=-1) -> Array:
```
[the remark this answers is not in the document]
Comment (ebezzam) [reply src/chromatix/utils/czt.py:46]: I didn't see any usage of `ArrayLike` elsewhere in the package, but it does seem like better practice to use that for inputs: https://docs.jax.dev/en/latest/jax.typing.html#jax-typing-best-practices

[the remark this answers is not in the document]
Comment (GJBoth) [reply src/chromatix/utils/czt.py:46]: Yes, because our typing relied on Chex when we did everything. Only after was it implemented in Jax, and I want to switch to that but haven't really gotten around to it.

File: pyproject.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (GJBoth) [reply pyproject.toml:16]: Fair, point. Should we move it to docs then?

File: src/chromatix/utils/czt.py (hunk 3 above)
Comment (ebezzam) [reply src/chromatix/utils/czt.py:16]: I'll look into checkify for some of the simpler checks. But won't get into jaxtyping since I don't see examples of it in the codebase and that would probably warrant a longer discussion...

File: src/chromatix/utils/czt.py (hunk 2 above)
Comment (ebezzam) [reply src/chromatix/utils/czt.py:90]: That's what the [scipy implementation](https://github.com/scipy/scipy/blob/a135e9763cca3e765f57ff9ea1b03775c8b82538/scipy/signal/_czt.py#L263) does! Probably where Claude got its inspiration 😉

File: pyproject.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ebezzam) [reply pyproject.toml:16]: It's not needed to built the docs though, so I would say `dev` is better?

File: src/chromatix/utils/czt.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    broadcast_shape[axis] = n
+    y = jnp.fft.fft(x * Awk2.reshape(broadcast_shape), n_czt, axis=axis)
+    broadcast_shape[axis] = n_czt
+    y = jnp.fft.ifft(y * Fwk2.reshape(broadcast_shape), axis=axis)
+    broadcast_shape[axis] = m
+    y = y[idx] * wk2.reshape(broadcast_shape)
+    return y
+
+
+def cztn(x: Array, m: List, a: List, w: List, axes: List) -> Array:
```
[the remark this answers is not in the document]
Comment (ebezzam) [reply src/chromatix/utils/czt.py]: Done

File: src/chromatix/utils/czt.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        Using float32/complex64 may have numerical accuracy issues.
+
+    Args:
+        x: Input signal to transform.
+        m: Number of samples in the output. List for each dimension.
+        a: The starting point in the complex plane. List for each dimension.
+        w: The ratio between points in each step. List for each dimension.
+        axes: Axes along which to perform the CZT.
+    """
+    x_czt = x.copy()
```
[the remark this answers is not in the document]
Comment (ebezzam) [reply src/chromatix/utils/czt.py]: Removed

File: src/chromatix/utils/czt.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    for d, ax in enumerate(axes):
+        x_czt = czt(x_czt, a=a[d], w=w[d], m=m[d], axis=ax)
+    return x_czt
+
+
+def _cztn(x: Array, m: List, a: List, w: List, axes: List) -> Array:
```
[the remark this answers is not in the document]
Comment (ebezzam) [reply src/chromatix/utils/czt.py]: Removed

File: src/chromatix/utils/czt.py (hunk 1 above)
Comment (ebezzam) [reply src/chromatix/utils/czt.py]: It isn't, I've removed and added some `checkify` checks

File: src/chromatix/utils/czt.py (hunk 2 above)
Comment (ebezzam) [reply src/chromatix/utils/czt.py]: I've changed to `moveaxis` and agree that it's much clearer

File: src/chromatix/utils/czt.py (hunk 3 above)
Comment (ebezzam) [reply src/chromatix/utils/czt.py]: I added `checkify` checks for `m` and `axis`. For checking that `a` and `w` are close to the unit circle, I couldn't see a clean way to do what with `checkify` (equality checks won't work because due to numerical errors `abs(w)` is not always exactly 1). So I've added in the docstrings that they should be on the unit circle.

Comment (GJBoth) [reply src/chromatix/utils/czt.py]: So I just checked the checkify docs and we'd need to add the transformation as well, and then handle the error. So it's a pretty complicated move. Jaxtyping does do checks during compile time, so I think I'd rather go with that. Should we just remove the check for now (commenting out is fine, with a note), and then look into jaxtyping and do this properly once and for all? WDYT?

**Comment by GJBoth** [comment]
Final thing: can you add test to see if `czt` and `cztn` run under `jit`? Once that's done let's merge it!

**Comment by ebezzam** [comment]
@GJBoth So original code wasn't compatible with `jit` because this [`arange`](https://github.com/ebezzam/chromatix/blob/d79957d32122aeb339b4d91afe8b7c4b82c0bccb/src/chromatix/utils/czt.py#L46) was a dynamic array, so I added an additional argument to pass the [input length](https://github.com/ebezzam/chromatix/blob/d79957d32122aeb339b4d91afe8b7c4b82c0bccb/src/chromatix/utils/czt.py#L27), which needs to be set as a [static argument](https://github.com/ebezzam/chromatix/blob/d79957d32122aeb339b4d91afe8b7c4b82c0bccb/tests/test_czt.py#L64).

Let me know if there's a better way to do it!

**Comment by GJBoth** [comment]
Exactly why I asked for the test :-). Jax can be a bit finicky re shapes. Static shapes using jit is indeed the way to go, but in practice a user will probably never need to do this; you jit at the highest level of your code at which this is all static.

Unless I'm missing something, I think this is ready to be merged?

**Comment by ebezzam** [comment]
@GJBoth yes all good now! Just did some last checks on `jit` and actually didn't need to set `n` at the input so I removed it.
